core/shared/src/main/scala/spire/math/interval/Overlap.scala

```diff
@@ -0,0 +1,90 @@
+package spire.math.interval
+
+import spire.algebra.{Eq, Order}
+import spire.math.Interval
+
+/**
+  * An ADT that represents overlapping result for any two intervals.
+  * All subclasses and method names are "left-biased":
+  * for example, [[StrictlyLess]] result means, that [[lhs]] is strictly less than [[rhs]].
+  */
+sealed abstract class Overlap[A: Order] extends Serializable {
+  def lhs: Interval[A]
+  def rhs: Interval[A]
+  def isStrictlyLess: Boolean
+  def isLessAndOverlaps: Boolean
+  def isSubset: Boolean
+  def isEqual: Boolean
+}
+
+/**
+  * Intervals are nonEmpty and don't intersect
+  * [[lhs.upperBound]] is strictly less than [[rhs.lowerBound]].
+  */
+case class StrictlyLess[A: Order] private[spire](lhs: Interval[A], rhs: Interval[A]) extends Overlap[A] {
```

Comment:
Could we move these case classes inside the companion object of `Overlap`? That seems to be the prevailing style now, even though we've been inconsistent in Spire historically.